src/webui/webapplication.cpp

```diff
@@ -838,153 +855,598 @@ void WebApplication::action_command_setLocation()
+        if (isInvalid)
+            LogMsg(tr("WebUI: Origin header & Target origin mismatch! Source IP: '%1'. Origin header: '%2'. Target origin: '%3'")
+                   .arg(m_env.clientAddress.toString()).arg(originValue).arg(targetOrigin)
+                   , Log::WARNING);
+        return isInvalid;
+    }
+
+    if (!refererValue.isEmpty()) {
+        const bool isInvalid = !isSameOrigin(urlFromHostHeader(targetOrigin), refererValue);
+        if (isInvalid)
+            LogMsg(tr("WebUI: Referer header & Target origin mismatch! Source IP: '%1'. Referer header: '%2'. Target origin: '%3'")
+                   .arg(m_env.clientAddress.toString()).arg(refererValue).arg(targetOrigin)
+                   , Log::WARNING);
+        return isInvalid;
+    }
+
+    return true;
+}
+
+bool WebApplication::validateHostHeader(const QStringList &domains) const
+{
+    const QUrl hostHeader = urlFromHostHeader(m_request.headers[Http::HEADER_HOST]);
+    const QString requestHost = hostHeader.host();
+
+    // (if present) try matching host header's port with local port
+    const int requestPort = hostHeader.port();
+    if ((requestPort != -1) && (m_env.localPort != requestPort)) {
+        LogMsg(tr("WebUI: Invalid Host header, port mismatch. Request source IP: '%1'. Server port: '%2'. Received Host header: '%3'")
+               .arg(m_env.clientAddress.toString()).arg(m_env.localPort)
+               .arg(m_request.headers[Http::HEADER_HOST])
+                , Log::WARNING);
+        return false;
+    }
+
+    // try matching host header with local address
+#if (QT_VERSION >= QT_VERSION_CHECK(5, 8, 0))
+    const bool sameAddr = m_env.localAddress.isEqual(QHostAddress(requestHost));
+#else
+    const auto equal = [](const Q_IPV6ADDR &l, const Q_IPV6ADDR &r) -> bool
+    {
+        for (int i = 0; i < 16; ++i) {
+            if (l[i] != r[i])
+                return false;
+        }
+        return true;
+    };
+    const bool sameAddr = equal(m_env.localAddress.toIPv6Address(), QHostAddress(requestHost).toIPv6Address());
+#endif
+
+    if (sameAddr)
+        return true;
+
+    // try matching host header with domain list
+    for (const auto &domain : domains) {
+        QRegExp domainRegex(domain, Qt::CaseInsensitive, QRegExp::Wildcard);
+        if (requestHost.contains(domainRegex))
+            return true;
+    }
+
+    LogMsg(tr("WebUI: Invalid Host header. Request source IP: '%1'. Received Host header: '%2'")
+           .arg(m_env.clientAddress.toString()).arg(m_request.headers[Http::HEADER_HOST])
+            , Log::WARNING);
+    return false;
+}
+
+void WebApplication::checkParams(const QSet<QString> &requiredParams, const QSet<QString> &optionalParams) const
+{
+    const QSet<QString> params {this->params().keys().toSet()};
+
+    if (!params.contains(requiredParams)
+            || ((request().method == QLatin1String("POST"))
+                && !optionalParams.contains(params - requiredParams)))
```

Comment:
`optionalParams`: what does this variable mean? if it's optional then it shouldn't throw any error at all (when not exists).
Maybe omit this field and let the respective function/method handle its existence?

Reply:
The meaning of this check: all required params must be presented, any of optional param may or may not be presented, any other params are forbidden (except the GET request, where any extra params are allowed).

Reply:
>any other params are forbidden 

Isn't that a bit too strict? IMO we should be tolerant about it.

Reply:
I think the same way. We shouldn't check it at all and just take the params we need. Initially there was no checks. But someone add it for some reason (I don't remember who done it).
In this PR I just allow optional params in addition to required.

Reply:
>Initially there was no checks. But someone add it for some reason 

I just think checking the **required** params is needed and is a common pattern (not sure this can be small enough to inline into each function), however the **optional** params shouldn't be checked at all.
What do you think?

Reply:
The existing code allows no unexpected params, I only implemented optional params as close as possible. But, as I said, this strictness not so important to me. Maybe leave it for separate issue?

Reply:
>Maybe leave it for separate issue?

OK.